### PR TITLE
resizing to align blocksize 

### PR DESCRIPTION
**Pull request by meiravgri**
### 1. in HNSW index
   - add: When exceeding capacity, it is increased and aligned with vector block size
   - delete: if we free an entire vector block, and the size is at least one block size smaller than the capacity, the capacity is 
     aligned to block size and decreased by block size.
   - check if label exist in hnsw_wrapper, before calling hnswlib add or delete
### 2. in brute force index:
   - add: When exceeding capacity (no more empty cells in idToVectorMemeberMapping) , it is increased and aligned with 
     vector block size
   - no resizing down upon deleteion because in the current logic, idToVectorMemeberMapping is not continous.

### CODE REVIEW

File: tests/unit/test_hnswlib.cpp [hunk 1]
```diff
@@ -54,19 +56,129 @@ TEST_F(HNSWLibTest, resizeIndex) {
...
+    ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(), n);
+
+    // add another vector to exceed the initial capacity
+    VecSimIndex_AddVector(index, (const void *)a, n);
+
+    // the capacity should be now aligned with the block size
+    ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity() % bs, 0);
+    // new capacity = intiailcapacity + blockSize - intiailcapacity % blockSize
+    ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(),
+              n + bs - n % bs);
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:80]: Consider testing the exact expected numbers

File: src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp [hunk 2]
```diff
@@ -112,8 +114,19 @@ int HNSWIndex::addVector(const void *vector_data, size_t id) {
...
     bool res = this->hnsw->removePoint(id);
-    if (hnsw->getIndexSize() + this->blockSize <= this->hnsw->getIndexCapacity()) {
-        this->hnsw->resizeIndex(this->hnsw->getIndexCapacity() - this->blockSize);
+    size_t index_size = hnsw->getIndexSize();
+    size_t block_size = this->blockSize;
+    size_t curr_capacity = this->hnsw->getIndexCapacity();
+
+    // if we need to free a complete block & there is a least one block between the
+    // capacity and the size
+    if (index_size % block_size == 0 && index_size + block_size <= curr_capacity) {
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp:123]: this can happen if the capacity is at least twice the block size,
if the condition will be 
index_size % block_size != 0 && index_size + block_size <= curr_capacity
each deleteion will cause the resizing of the index, until the gap is reduced

File: tests/unit/test_hnswlib.cpp [hunk 3]
```diff
@@ -39,11 +39,13 @@ TEST_F(HNSWLibTest, hnswlib_vector_add_test) {
     ASSERT_EQ(VecSimIndex_IndexSize(index), 1);
     VecSimIndex_Free(index);
 }
+/**** resizing cases ****/
 
-TEST_F(HNSWLibTest, resizeIndex) {
+// add up to capacity
+TEST_F(HNSWLibTest, resizeNAlignIndex) {
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_hnswlib.cpp:45]: :)

File: tests/unit/test_hnswlib.cpp [hunk 4]
```diff
@@ -54,19 +56,129 @@ TEST_F(HNSWLibTest, resizeIndex) {
...
+
+    // add up to n
+    for (size_t i = 0; i < n; i++) {
+        for (size_t j = 0; j < dim; j++) {
+            a[j] = (float)i;
+        }
+        VecSimIndex_AddVector(index, (const void *)a, i);
+    }
+    // the size and the capacity should be equal
+    ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(),
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_hnswlib.cpp:68]: this cast was in the original test, i agree that it is not the best voding style
but i need the capacity not the size 

maybe we can add VecSimIndex_NewHNSW ?

File: tests/unit/test_hnswlib.cpp [hunk 5]
```diff
@@ -54,19 +56,129 @@ TEST_F(HNSWLibTest, resizeIndex) {
...
+            a[j] = (float)i;
+        }
+        VecSimIndex_AddVector(index, (const void *)a, i);
+    }
+
+    // the capcity shouldnt be changed
+    ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(), n);
+
+    // delete random vector, to get size % block_size == 0
+    VecSimIndex_DeleteVector(index, 1);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_hnswlib.cpp:113]: no. and thats the wanted behaviour

File: src/VecSim/algorithms/brute_force/brute_force.cpp [hunk 6]
```diff
@@ -85,16 +85,22 @@ int BruteForceIndex::addVector(const void *vector_data, size_t label) {
         if (this->deletedIds.size() != 0) {
             id = *this->deletedIds.begin();
             this->deletedIds.erase(this->deletedIds.begin());
-            this->count++;
         } else {
-            id = this->count++;
+            id = count;
+            // save current id2vec size
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force.cpp:90]: revisited the files and fixed it, hope i didnt miss anything

File: src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp [hunk 7]
```diff
@@ -112,8 +114,19 @@ int HNSWIndex::addVector(const void *vector_data, size_t id) {
 
 int HNSWIndex::deleteVector(size_t id) {
     bool res = this->hnsw->removePoint(id);
-    if (hnsw->getIndexSize() + this->blockSize <= this->hnsw->getIndexCapacity()) {
-        this->hnsw->resizeIndex(this->hnsw->getIndexCapacity() - this->blockSize);
+    size_t index_size = hnsw->getIndexSize();
+    size_t block_size = this->blockSize;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp]: you are very right

File: src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp]: I understand now that this is intentional. Sounds good

File: tests/unit/test_hnswlib.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/unit/test_hnswlib.cpp:68]: For the capacity you should use `reinterpret_cast` as you did, but when you check for the index size you can use `VecSimIndex_IndexSize` - it's cleaner...

File: tests/unit/test_hnswlib.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/unit/test_hnswlib.cpp]: I understand. Do you have a test case for this scenario (deleting a vector when there is a block to remove, but avoiding it because `size % block_size != 0`). If not, consider adding it.

File: src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp [hunk 8]
```diff
@@ -113,22 +119,29 @@ int HNSWIndex::addVector(const void *vector_data, size_t id) {
...
 
 int HNSWIndex::deleteVector(size_t id) {
-    bool res = this->hnsw->removePoint(id);
+
+    // If id doesnt exist.
+    if (!this->hnsw->isLabelExist(id)) {
+        return false;
+    }
+
+    // Else delte it from the graph.
```
Comment (alonre24) [inline src/VecSim/algorithms/hnsw/hnsw_wrapper.cpp:128]:
```suggestion
    // Else, *delete* it from the graph.
```

File: tests/unit/test_bruteforce.cpp [hunk 9]
```diff
@@ -57,14 +57,44 @@ TEST_F(BruteForceTest, resizeIndex) {
...
+
+    // Size should be n + 1 + 8 = 24.
+    ASSERT_EQ(VecSimIndex_IndexSize(index), n + 1 + add_vectors_count);
+    // Check alignment of the capacity
+    ASSERT_EQ(reinterpret_cast<BruteForceIndex *>(index)->idToVectorBlockMemberMapping.size() %
+                  blockSize,
+              0);
+    // Check new capacity size, should be blockSize * 3.
+    ASSERT_EQ(reinterpret_cast<BruteForceIndex *>(index)->idToVectorBlockMemberMapping.size(),
+              3 * blockSize);
```
Comment (alonre24) [inline tests/unit/test_bruteforce.cpp:98]: The test in line 92 is redundant, you can leave only the one in line 96

File: tests/unit/test_hnswlib.cpp [hunk 10]
```diff
@@ -57,32 +57,34 @@ TEST_F(HNSWLibTest, resizeNAlignIndex) {
...
+    // The capacity should be now aligned with the block size.
+    // bs = 3, size = 11 -> capacity = 12
     ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity() % bs, 0);
-    // new capacity = intiailcapacity + blockSize - intiailcapacity % blockSize
+    // New capacity = intiailcapacity + blockSize - intiailcapacity % blockSize.
     ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(),
               n + bs - n % bs);
-
+    // Check actual numbers **cahnge this if you change the intial parameters**
+    ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(), 12);
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:83]: You have some redundancy here, you can remove the tests in lines 78 and 83

File: tests/unit/test_hnswlib.cpp [hunk 11]
```diff
@@ -57,32 +57,34 @@ TEST_F(HNSWLibTest, resizeNAlignIndex) {
...
     VecSimIndex_Free(index);
 }
 
-// case 1: intiailCapacity much larger than block size, and it is not aligned
+// Case 1: intiailCapacity much larger than block size, and it is not aligned.
 TEST_F(HNSWLibTest, resizeNAlignIndex_initailCapacity_is_big) {
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:88]: Typo: initial capacity. Also, in comments I recommend writing English words rather than variable names as much as possible...

Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:88]:
```suggestion
TEST_F(HNSWLibTest, resizeNAlignIndex_largeInitialCapacity) {
```

File: tests/unit/test_hnswlib.cpp [hunk 12]
```diff
@@ -98,44 +100,60 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_initailCapacity_is_big) {
...
+    // add up to blocksize + 1 = 3 + 1 = 4
     for (size_t i = 0; i < bs + 1; i++) {
         for (size_t j = 0; j < dim; j++) {
             a[j] = (float)i;
         }
         VecSimIndex_AddVector(index, (const void *)a, i);
     }
 
-    // the capcity shouldnt be changed
+    // The capcity shouldnt be changed. (32)
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:111]:
```suggestion
    // The index capacity shouldn't change (should remain `n`)
```

File: tests/unit/test_hnswlib.cpp [hunk 13]
```diff
@@ -98,44 +100,60 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_initailCapacity_is_big) {
...
     ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(), n);
 
-    // delete random vector, to get size % block_size == 0
+    // Delete random vector, to get size % block_size == 0. size = 3
     VecSimIndex_DeleteVector(index, 1);
 
-    ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexSize(), bs);
+    // Index size = bs = 3.
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:117]: Why did you remove the test itself? 
I would add here:
`ASSERT_EQ(VecSimIndex_IndexSize(index), bs)`
instead of the comment

File: tests/unit/test_hnswlib.cpp [hunk 14]
```diff
@@ -98,44 +100,60 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_initailCapacity_is_big) {
...
+    ASSERT_EQ(VecSimIndex_IndexSize(index), bs);
 
-    // the capacity should be now aligned with the block size
+    // The capacity should be now aligned with the block size.
     ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity() % bs, 0);
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:121]: redundant (you have line 124)

File: tests/unit/test_hnswlib.cpp [hunk 15]
```diff
@@ -98,44 +100,60 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_initailCapacity_is_big) {
...
-    // new capacity = intiailcapacity - block_size - number_of_vectors_to_align
+    // New capacity = intiailcapacity - block_size - number_of_vectors_to_align =
+    // 32  - 3 - 32 % 3 (2) = 27
     ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(),
               n - bs - n % bs);
+    // Check actual numbers **cahnge this if you change the intial parameters**
+    ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(), 27);
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:127]: redundant (you have line 124)

File: tests/unit/test_hnswlib.cpp [hunk 16]
```diff
@@ -98,44 +100,60 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_initailCapacity_is_big) {
...
 
     size_t curr_capacity = reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity();
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:129]: To have cleaner code, you can move this to be before line 124, and test instead:
`ASSERT_EQ(curr_capacity, n - bs - n % bs);`

File: tests/unit/test_hnswlib.cpp [hunk 17]
```diff
@@ -98,44 +100,60 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_initailCapacity_is_big) {
...
-    // add another vector - > the capacity remains the same
+    // Add another vector - > the capacity remains the same, size = 4.
     VecSimIndex_AddVector(index, (const void *)a, n);
     ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(),
               curr_capacity);
 
-    // delete the vector -> the capacity is decreased by block size
-    VecSimIndex_DeleteVector(index, bs);
+    // Delete the vector -> the capacity is again decreased by block size.
+    VecSimIndex_DeleteVector(index, 2);
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:136]: redundant (you have line 138)

File: tests/unit/test_hnswlib.cpp [hunk 18]
```diff
@@ -98,44 +100,60 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_initailCapacity_is_big) {
...
 
     ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(),
               curr_capacity - bs);
+    // Check actual numbers **cahnge this if you change the intial parameters**
+    ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(), 24);
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:141]: redundant

File: tests/unit/test_hnswlib.cpp [hunk 19]
```diff
@@ -98,44 +100,60 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_initailCapacity_is_big) {
...
+
+    curr_capacity = reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity();
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:143]: Here also, you can remove this to be before line 138

File: tests/unit/test_hnswlib.cpp [hunk 20]
```diff
@@ -98,44 +100,60 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_initailCapacity_is_big) {
...
+
+    // Delete another vector
+    VecSimIndex_DeleteVector(index, 3);
+
+    // Size + bs <= capacity (2 + 3 <= 24), but size %bs != 0.
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:148]: typo: capacity.
Also, mising space after the `%` (sorry for being petty :))

File: tests/unit/test_hnswlib.cpp [hunk 21]
```diff
@@ -98,44 +100,60 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_initailCapacity_is_big) {
...
+    // capapcity should remain the same.
+    ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity(),
+              curr_capacity);
 
     VecSimIndex_Free(index);
 }
 
-// case 2: intiailCapacity smaller than block_size
+// Case 2: intiailCapacity smaller than block_size.
 TEST_F(HNSWLibTest, resizeNAlignIndex_blocksize_is_bigger) {
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:157]:
```suggestion
TEST_F(HNSWLibTest, resizeNAlignIndex_largerBlocSize) {
```

File: tests/unit/test_hnswlib.cpp [hunk 22]
```diff
@@ -151,37 +169,81 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_blocksize_is_bigger) {
...
+    // Add up to initial capacity.
     for (size_t i = 0; i < n; i++) {
         for (size_t j = 0; j < dim; j++) {
             a[j] = (float)i;
         }
         VecSimIndex_AddVector(index, (const void *)a, i);
     }
 
-    // the capcity shouldnt be changed
+    // The capcity shouldnt be changed.
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:180]:
```suggestion
    // The capacity shouldn't change.
```

File: tests/unit/test_hnswlib.cpp [hunk 23]
```diff
@@ -151,37 +169,81 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_blocksize_is_bigger) {
...
 
-    // size equals capacity
-    ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexSize(), n);
+    // Size equals capacity.
+    ASSERT_EQ(VecSimIndex_IndexSize(index), n);
 
-    // add another vector - > the capacity is increased to a multiplication of block_size
+    // Add another vector - > the capacity is increased to a multiplication of block_size.
     VecSimIndex_AddVector(index, (const void *)a, n);
     ASSERT_EQ(reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity() % bs, 0);
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:188]: redundant (you have line 189)

File: tests/unit/test_hnswlib.cpp [hunk 24]
```diff
@@ -151,37 +169,81 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_blocksize_is_bigger) {
...
+    // Add one vector.
+    float a[dim];
+    for (size_t j = 0; j < dim; j++) {
+        a[j] = (float)1.7;
+    }
+
+    VecSimIndex_AddVector(index, (const void *)a, 1);
+    // Try to remove it.
+    VecSimIndex_DeleteVector(index, 1);
+    // The capcity should be changed and aligned to vector size.
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:226]:
```suggestion
    // The capacity should change to be aligned with the vector size.
```

File: tests/unit/test_hnswlib.cpp [hunk 25]
```diff
@@ -151,37 +169,81 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_blocksize_is_bigger) {
...
+
+    size_t new_capcaity = reinterpret_cast<HNSWIndex *>(index)->getHNSWIndex()->getIndexCapacity();
+    ASSERT_EQ(new_capcaity, n - n % bs - bs);
+    // Sanity check ***hardcodeed numbers****.
+    ASSERT_EQ(new_capcaity, 12);
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:231]: redundant...

**Review by alonre24 — COMMENTED** [review]
Last comments

File: tests/unit/test_hnswlib.cpp (hunk 13 above)
Comment (meiravgri) [reply tests/unit/test_hnswlib.cpp:117]: its in line 118?

Comment (alonre24) [reply tests/unit/test_hnswlib.cpp:114]: right! sorry :)

File: tests/unit/test_hnswlib.cpp [hunk 26]
```diff
@@ -177,15 +167,18 @@ TEST_F(HNSWLibTest, resizeNAlignIndex_blocksize_is_bigger) {
         VecSimIndex_AddVector(index, (const void *)a, i);
     }
 
-    // The capcity shouldnt be changed.
+<<<<<<< HEAD
+    // The capcity shouldnt change.
+=======
+    // The capacity shouldn't change.
+>>>>>>> 394e7922ab865eb1ad418eac3bf7d3a6a8636362
```
Comment (alonre24) [inline tests/unit/test_hnswlib.cpp:174]: You have a leftover here...

**Review by alonre24 — APPROVED** [review]

**Review by DvirDukhan — CHANGES_REQUESTED** [review]
Great
1. Please edit the PR description to describe your changes
2. please add a test where capacity is dropped to 0. If it is already covered by one of the existing tests, just mention it.

**Review by DvirDukhan — APPROVED** [review]
